### ISSUE TITLE
expandRepeats is extremely slow

### ISSUE THREAD

**Issue by giamic** [cuthbertLab/music21#1502]
**music21 version**

`8.1.0` , tested also with the latest github master branch  

**Problem summary**

The code hangs when expanding repeats.

**Steps to reproduce**
Download https://github.com/MarkGotham/When-in-Rome/blob/master/Corpus/Piano_Sonatas/Mozart%2C_Wolfgang_Amadeus/K284/3/score.mxl , then run this script

```
from music21 import converter

fp = "path/to/the/K284.3/score.mxl"
score = converter.parse(fp)
score.expandRepeats()
```

**Expected vs. actual behavior**

The code hangs after this call. The expected behaviour is to obtain a version of the score with expanded repeats in a short time.

**More information**

On Apple M1, MacOS 12.6.

Profiling shows that, after 409 seconds, 99% of time is used by expandRepeats and 96.3% is occupied in deepcopies. There have been more than 78 million calls to deepcopy so far.

**Comment by giamic** [comment]
Update: There seems to be an error in the score itself for which some start repeat bars are missing which made the task more complex. Once fixed, `expandRepeats()` takes 37 seconds and does 11 million calls to deepcopy. This allowed me to finally process the file but doesn't fundamentally matter for this issue.

**Comment by MarkGotham** [comment]
Thanks @giamic. Sounds like the main dev needed here is for `expandRepeats()` raise error and reject a score like this as badly formed.

That would mean a short routine at the start of `expandRepeats()` that check validity. (Or if there's already such a routine then fix/expand to catch cases like this).

What does "validity" mean? It's not quite so simple as making sure repeat marks are paired. Off the top of my head, how's this logic based on a list of start and end repeats:
- For every end repeat _after the first one_ `[1:]`,
- There must be a start repeat _since the previous end repeat_,

No check on the first end repeat because that can safely do without a start repeat (back to start of score).

That would at least fix the issue @giamic found in this DCML Mozart score: missing start repeats at the beginning of each variation. E.g., see such an omission at the beginning of `Var. I` here:
<img width="520" alt="DCML_eg" src="https://user-images.githubusercontent.com/19486660/207626931-8d3314b4-3c81-4dc4-94f9-bc1c11699ad7.png">

**Comment by gregchapman-dev** [comment]
From the peanut gallery: I would think that this proposed "validateRepeats" routine would be valuable in its own right. Please don't hide it.

**Comment by mscuthbert** [comment]
There are nested repeats. So we shouldn't reject on that. So it's not that it needs a repeat after the following end. It's that each repeat end needs a unique repeat start (including the implied start at the beginning of the piece)

**Comment by MarkGotham** [comment]
@mscuthbert are you talking about _da capo_ blocks? Surely not nested double bar repeat marks like
`||: notes ||: notes :|| notes :||`?

When in Rome: Over there I expect that the quick and easy test that repeats match will be plenty, so I'll implement that and check it covers everything we encounter. I'll refine it at bit (e.g., it can't currently return False), but this works at least on the issue that @giamic raises, shown above (catches the first such problem as shown):

```
def repeats_are_valid(pathToScore: str) -> bool:
    """
    Quick and simple check that start and end repeats match up.
    Specifically, iterates through the measures on the highest part
    and raises a value error in the case of:
    two starts without and end,
    or two ends without a start.
    
    Prints a log of every repeat it finds,
    raises an error in the case of an issue,
    returns True if no such issues.
    """
    score = converter.parse(pathToScore)
    last_repeat = ''  # starting condition
    for measure in score.parts[0].recurse().getElementsByClass(stream.Measure):
        if measure.leftBarline:
            if last_repeat == 'Start':
                # Two starts! Time to end!
                raise ValueError(f'Second successive start repeat found in measure {measure.measureNumber}.')
            else:
                print(f'Start repeat in measure {measure.measureNumber}')
                last_repeat = 'Start'
        if measure.rightBarline:
            if last_repeat == 'End':
                # Two ends! Time to stop!
                raise ValueError(f'Second successive end repeat found in measure {measure.measureNumber}.')
            else:
                print(f'End repeat in measure {measure.measureNumber}')
                last_repeat = 'End'
    
    return True
```

@giamic - does that satisfy you?

@mscuthbert any prospective music21 integration is up to you. E.g., you're welcome to
- request a PR of this
- request something more elaborate/specific
- wait to see we encounter further issues in practice
- ignore the whole thing and close the issue!

As always: open offer, no pressure or expectation.
And in December: Greeting of the seasons to all! Ho ho ho!

**Comment by giamic** [comment]
@MarkGotham I haven't tried it on my machine but the logic looks good. I propose a few changes, namely:
- accept a score instead of a path in input: reading the score is not a logical component of the check and this can save us time in case the score has already been parsed (and I imagine that would be the case most often)
- iterate through the barlines instead of the measures: tests show that this is not faster then iterating through the measures, but it simplifies the code and make it easier to read for humans
- remove the prints at every valid repeat (it was flooding the stdout in cases with many repeats and that feels like an unnecessary information)
- do not return True: I feel it's counterintuitive to have an error if there is a problem and True if everything is hunky dory — it's like mixing two different and perfectly equivalent style

The final result is the following
```python
def repeats_are_valid(score: stream.Score) -> None:
    """
    Quick and simple check that start and end repeats match up.
    Specifically, iterates through the barlines on the highest part
    and raises a value error in the case of:
    two starts without an end,
    or two ends without a start.
    """
    last_repeat = ""  # starting condition
    for barline in score.parts[0].recurse().getElementsByClass(bar.Barline):
        if str(barline) == str(bar.Repeat(direction="start")):
            if last_repeat == "start":
                raise ValueError(
                    f"Second successive start repeat found in measure {barline.measureNumber}."
                )
            else:
                last_repeat = "start"
        if str(barline) == str(bar.Repeat(direction="end")):
            if last_repeat == "end":
                raise ValueError(
                    f"Second successive end repeat found in measure {barline.measureNumber}."
                )
            else:
                last_repeat = "end"
    return
```
Anyway, they are kind of minor changes to your code, thank you very much for the solution!

Ho ho ho to you! :D

**Comment by mscuthbert** [comment]
> @mscuthbert are you talking about _da capo_ blocks? Surely not nested double bar repeat marks like `||: notes ||: notes :|| notes :||`?

Yes.  These appear often both in Broadway musicals and in minimalist compositions.  Generally the outer repeats have bigger wings to distinguish them.

Note also that the S-shaped repeat sign is also a repeat mark, so keeping track of opened and closed repeats is still necessary.

It just means working with Catalan numbers (Dyck Words) -- it's not too hard, but does require an O(n) scan through the stream that can't be short circuited.

**Comment by MarkGotham** [comment]
Interesting, thanks! I'm obviously not spending enough time with Broadway/minimalist scores!

Some thoughts, FWIW:

Wrt S-shaped repeat signs, da capos and the like, MuseScore has a nice solution to this in which repeat markers indicate multiple positions:
- "Jump to" (the next measure to be played) as a well effectively the next jump pair:
- "Play until" (the next "jump from")
- "Continue at" (the next "jump to").

That could be a useful model.

Wrt nested repeat marks. I'm sure it's soluble, and I'm all for supporting variant ways of encoding music, but is this nesting practice one we want to perpetuate? Is it defined and supported formally in musicXML and the like? I'd say:
- if there's differentiation then great: get the relevant attribute, (e.g. `large` and `small` repeat labels or `repeatLayer=1`) and require pairs of them to match (`large start` to `large end`).
- if not, I'd think that nested repeats _without such differentiation_ is an inherently problematic convention for both print and digital scores, and beyond what's reasonable to perpetuate by encoding it here. (Not possible for music21 to support everything, of course, so this is hardly a one-off). I was asking around about this and one person came up with a similar(ish) case of "petite reprise" in Couperin. Quite independently of this discussion, they considered this outside of standard notation practice now and wrote the thing out in the new edition.

Up to you to define the limits and philosophy of music21, of course. I dare say that this author of this issue (@giamic) has lost interest, and I'm I'm satisfied with something like the simply solution above for the purposes of WiR.

Seasonal greetings to all.

**Comment by giamic** [comment]
I did not lose interest in this issue. The issue originally was related to the speed of expandRepeats() which tends to be extremely slow, with the majority of time taken by an incredibly large amount of deepcopies. I'm still interested in how to improve the performance of music21.

However, over the course of the discussion, the topic has slowly shifted to "how to define proper repeat bars" which is a very interesting topic in itself but one which I don't feel I can contribute to.

**Comment by jacobtylerwalls** [comment]
@giamic One track to explore could be adding an `inPlace=True` argument to expandRepeats (and perhaps adjusting the behavior of Expander.expand() when deepcopy=False). I think the deep copies are made because expansion has the potential to insert the same object into a stream hierarchy twice, which is not allowed in music21. But that's only relevant for the measures being duplicated. You could imagine a compromise where the passages not being duplicated don't need deep copies, and we just deep copy what's necessary.

Would you be interested to evaluate the feasibility of something like that?

**Comment by giamic** [comment]
Hi @jacobtylerwalls , that is a good explanation and it sounds like a very good idea to try. Unfortunately I have moved to the next project which does not use music21, so at the minute I don't have the bandwidth to look into it. I will let you know if that changes in the future!
